### PR TITLE
Enable agent to execute code

### PR DESCRIPTION
**Pull request by fivegrant**
This PR provides a `run_code` tool to any agent using a checkpointable kernel. The specific changes made by this PR:
- Agent inherits tools from specific subkernels. This means tools can now be specified for a language as opposed to a context.
- The new disable functionality replaces `toggeable_tool`. **_NOTE: This is a breaking change since a few contexts use `toggleable_tool` in askem-beaker. However, the contexts using this are erroneously reimplementing `ask_user` anyway._**
- Add a new `run_code` tool implemented as a subkernel tool on CheckpointableSubkernel

## PRE-MERGE CHECKLIST
- [x] [Merge the disable functionality into `main` branch](https://github.com/jataware/archytas/pull/29)
- [x] Cut and publish a new release of archytas
- [x] Update `pyproject.toml` to use the new version

### CODE REVIEW

File: beaker_kernel/lib/subkernels/base.py [hunk 1]
```diff
@@ -128,3 +130,32 @@ def cleanup(self):
...
+        Args:
+            code (str): Code to run directly in Jupyter.
+        Returns:
+            str: Result of the `expr`
+        
+        """
+        checkpoint_index = await self.add_checkpoint()
+        result = await self.evaluate(code)
+        await self.rollback(checkpoint_index)
+        return result
```
Comment (mattprintz) [inline beaker_kernel/lib/subkernels/base.py:161]: Unless I'm mistaken, `result` will not be a string.

File: beaker_kernel/lib/utils.py [hunk 2]
```diff
@@ -164,18 +164,6 @@ def register_magic(fn):
     return register_magic
 
 
-def togglable_tool(env_var, *, name: str | None = None):
```
Comment (mattprintz) [inline beaker_kernel/lib/utils.py:167]: If you remove this here, it will break existing contexts outside of this repo.
Any reason to not just leave it in as deprecated?

File: beaker_kernel/lib/subkernels/base.py [hunk 3]
```diff
@@ -128,3 +130,32 @@ def cleanup(self):
         if self.checkpoints_enabled:
             shutil.rmtree(self.storage_prefix, ignore_errors=True)
             self.checkpoints = []
+
+
+    @tool()
```
Comment (mattprintz) [inline beaker_kernel/lib/subkernels/base.py:135]: Can you explain to me the thinking as to why it makes sense to define tools on using the @tool decorator on subkernel classes?
Previously, tools were only able to be defined as part of an agent and on a toolset class.
Keep in mind that the prompting for tools also takes in to consideration the parent class that an @tool is defined on.

File: beaker_kernel/lib/agent.py [hunk 4]
```diff
@@ -40,6 +40,11 @@ def __init__(
             thought_handler=context.beaker_kernel.handle_thoughts,
             **kwargs
         )
+        if not env_enabled("ENABLE_USER_PROMPT"):
```
Comment (mattprintz) [inline beaker_kernel/lib/agent.py:43]: By continuing to gate these behind an environment variable, it's not possible, or at least straightforward, to turn these on/off on a per-context basis.

File: beaker_kernel/lib/context.py [hunk 5]
```diff
@@ -41,12 +41,12 @@ def __init__(self, beaker_kernel: "LLMKernel", agent_cls: "BaseAgent", config: D
         self.jinja_env = None
         self.templates = {}
         self.beaker_kernel = beaker_kernel
+        self.config = config
+        self.subkernel = self.get_subkernel()
         self.agent = agent_cls(
             context=self,
-            tools=[],
+            tools=[self.subkernel],
```
Comment (mattprintz) [inline beaker_kernel/lib/context.py:48]: What happens to the prompting if you add a class that doesn't have any tools?

File: beaker_kernel/lib/subkernels/base.py (hunk 1 above)
Comment (fivegrant) [reply beaker_kernel/lib/subkernels/base.py:161]: https://github.com/jataware/beaker-kernel/blob/b8bd402d098a46f52167f38fb2987cd4076d7096/beaker_kernel/lib/context.py#L397-L405

https://github.com/jataware/beaker-kernel/blob/b8bd402d098a46f52167f38fb2987cd4076d7096/beaker_kernel/lib/subkernels/python.py#L52-L57

I think evaluate cleans it up for you

File: pyproject.toml [hunk 6]
```diff
@@ -30,7 +30,7 @@ dependencies = [
   "matplotlib~=3.7.1",
   "xarray==0.19.0",
   "numpy~=1.24.3",
-  "archytas~=1.1.5",
+  "archytas @ git+https://github.com/jataware/archytas.git@98e4ed87cc0b08a9b506151f1966f4117811f008",
```
Comment (mattprintz) [inline pyproject.toml:33]: Make sure you don't commit this.

File: beaker_kernel/lib/utils.py (hunk 2 above)
Comment (fivegrant) [reply beaker_kernel/lib/utils.py:167]: https://github.com/search?q=repo%3ADARPA-ASKEM%2Fbeaker-kernel+togglable_tool&type=code

The places where it's used should just be deleted. It's duplicating functionality that already exists in beaker-kernel

File: pyproject.toml (hunk 6 above)
Comment (fivegrant) [reply pyproject.toml:33]: don't worry, i put this on the pre-merge checklist :+1:

File: beaker_kernel/lib/context.py (hunk 5 above)
Comment (fivegrant) [reply beaker_kernel/lib/context.py:48]: https://github.com/jataware/archytas/blob/36c8ace7400b0b81abb24e7fa77a75289b6b0c5e/archytas/tool_utils.py#L41-L50

I'll double check, but if an empty class somehow breaks an issue needs to be opened up in archytas.

File: beaker_kernel/lib/subkernels/base.py (hunk 1 above)
Comment (mattprintz) [reply beaker_kernel/lib/subkernels/base.py:161]: I don't think so. Have you tested it?

File: beaker_kernel/lib/subkernels/base.py (hunk 3 above)
Comment (fivegrant) [reply beaker_kernel/lib/subkernels/base.py:135]: As I mentioned above in the PR:
> Agent inherits tools from specific subkernels. This means tools can now be specified for a language as opposed to a context.

[Toolsets are deprecated in archytas](https://github.com/jataware/archytas/blob/36c8ace7400b0b81abb24e7fa77a75289b6b0c5e/archytas/tool_utils.py#L41-L50). Any class can bundle together related functionality.

Currently, checkpointing only exists for Python so we only want to include the `run_code` tool when loading that specific subkernel. I've also in the past wanted to have some context-specific tools apply to all Julia contexts instead.

File: beaker_kernel/lib/agent.py (hunk 4 above)
Comment (fivegrant) [reply beaker_kernel/lib/agent.py:43]: Just to double check, would the idea here be that all the contexts in askem-beaker have these lines instead?

File: beaker_kernel/lib/utils.py (hunk 2 above)
Comment (fivegrant) [reply beaker_kernel/lib/utils.py:167]: If [the answer](https://github.com/jataware/beaker-kernel/pull/47#discussion_r1594569798) to this other question below is yes, we're going to have to make modifications to askem-beaker anyway

Comment (mattprintz) [reply beaker_kernel/lib/utils.py:167]: Perhaps, but you are making things harder on yourself by enforcing synchronicity on deployments that isn't needed.
If the other changes you made make this decorator not do anything then it's ok to remove the code and leave a warning.

```
import warnings
def togglable_tool(env_var, *, name: str | None = None):
    warnings.warn("This decorator is deprecated and will be removed in version 1.6.0")
    def noop(fn):
        return fn
    return noop
```

File: beaker_kernel/lib/context.py (hunk 5 above)
Comment (mattprintz) [reply beaker_kernel/lib/context.py:48]: I think it is pretty unlikely to break things, as in to cause errors, but I think that it could change the prompt that the archytas react agent generates.
Look in to how the prompt is generated around tools.

File: beaker_kernel/lib/subkernels/base.py (hunk 3 above)
Comment (mattprintz) [reply beaker_kernel/lib/subkernels/base.py:135]: Yes, I'm with you on defining tools that are defined depending on the kernel.

The thing that I'm getting at is that there are side-effects to how you are doing this, namely that it changes the prompt in ways I don't think you realize, and that there are alternative ways of doing what you want that would have less or no side-effects.
https://github.com/jataware/archytas/blob/main/archytas/tool_utils.py#L218

Namely: 
You can code this tool and similar tools as stand-alone functions and use the injections to get back to the agent/context for running code in the subkernel.

File: beaker_kernel/lib/context.py (hunk 5 above)
Comment (mattprintz) [reply beaker_kernel/lib/context.py:48]: When pulling the info for the comment I just left, I saw that there isn't a problem if there are no tools defined. 
i.e.:
```
        tool_methods = collect_tools_from_object(obj)
        if tool_methods:
          # All the stuff that changes the prompt
```

File: beaker_kernel/lib/subkernels/base.py (hunk 1 above)
Comment (fivegrant) [reply beaker_kernel/lib/subkernels/base.py:161]: Ah, I see. It was working with the LLM so I thought it was fine.

Just tested like this:
```
    @action()
    async def test_eval(self, message):
        """
        """
        a = await self.evaluate("1")
        b = await self.evaluate("'string'")
        c = await self.evaluate("['a', 'b']")
        return [str(type(a)), str(type(b)), str(type(c))]
```

I reread the above I somehow missed line 401. I'll change this to `"return"`. Will resolve in a future proof. Thanks for catching!

File: beaker_kernel/lib/agent.py (hunk 4 above)
Comment (mattprintz) [reply beaker_kernel/lib/agent.py:43]: The idea is that there would be a default and if different contexts wanted to deviate from the default there would be a way to do so.

The __init__ function allows **kwargs for exactly this sort of expansion. It would be a great method to allow the ability to override, but it could be a class attribute too. Maybe a combination of the two?

But no, the idea is not to have all the two lines in each context. For today, all the code should be in this file.

File: beaker_kernel/lib/subkernels/base.py (hunk 1 above)
Comment (mattprintz) [reply beaker_kernel/lib/subkernels/base.py:161]: On top of what you've discovered, keep in mind that the value of "return" will be the actual reified object, not a string. At some point down the line, it will be turned in to a string either via formatting or a direct call to str(return), but keep in mind if you `return result["return"]`, the type returned is `Any`, not `str`.

File: beaker_kernel/lib/context.py (hunk 5 above)
Comment (fivegrant) [reply beaker_kernel/lib/context.py:48]: I'm going to resolve this thread because I think it's very similar to https://github.com/jataware/beaker-kernel/pull/47#discussion_r1594587681

Feel free to reopen though

File: beaker_kernel/lib/subkernels/base.py (hunk 1 above)
Comment (mattprintz) [reply beaker_kernel/lib/subkernels/base.py:161]: Actually, as I was doing something else, I realized that the agent "wants" more than just the return value. It'll want the stdout and the stderr and probably other things too in addition to the return. Probably best to format it to a string formatted specifically for the agent to use, with headers and seperated sections, etc.

**Comment by fivegrant** [comment]
Example of the current performance:
image

File: beaker_kernel/lib/subkernels/base.py (hunk 3 above)
Comment (fivegrant) [reply beaker_kernel/lib/subkernels/base.py:135]: All right, I solved this by collecting the tools from the object before passing them to the agent.

I implemented the current solution after looking into moving the tool out of the subkernel class. Converting was simple enough:
```python
@tool()
async def run_code(code: str, agent: AgentRef) -> str:
        """
        Execute code in the user's session. After execution,
        the state of the kernel will be rolled back to before this tool
        was used.

        This tool can be help answer questions about the kernel state. For
        example, a user may ask something about a dictionary `d` and using 
        run code with the `code` of `d.keys()`.

        This tool can also be used to double check if code will work before 
        returning it as a final answer.

        Note that this tool does not capture `stdout` AND only returns the
        results of the last expression evaluated.

        Args:
            code (str): Code to run directly in Jupyter.
        Returns:
            str: Result of the `expr`
        
        """
        checkpoint_index = await agent.context.subkernel.add_checkpoint()
        result = await agent.context.evaluate(code)
        await agent.context.subkernel.rollback(checkpoint_index)
        return str(result["return"])
```
It feels a bit funky to use `agent.context.subkernel` twice. We're separating the code pretty far from all the methods it's trying to use directly. 

Compounding the above, it makes thinking through enabling a bit trickier. I initially thought that the user could just pass `run_code` into context initialization if they wanted to use it. However, it gets tricky when they make a context that supports both a checkpointable and non-checkpointable kernel. Then, we'll need some other kind of logic to disable the tool when the non-checkpointable tool is used. We could add refs to the available tools in the subkernel class but that seems less ideal than what we have above. We could also just check using the language string although I think the current implementation (now that I've removed the class from the prompt) is succinct and intuitive. [Note this does not solve the more general questions about enabling](https://github.com/jataware/beaker-kernel/pull/47#discussion_r1594546342)

File: beaker_kernel/lib/agent.py (hunk 4 above)
Comment (fivegrant) [reply beaker_kernel/lib/agent.py]: Inspired by your suggestion [here](https://jataware.slack.com/archives/C03J0RSKFK8/p1715262724438199?thread_ts=1715196514.594459&cid=C03J0RSKFK8), I implemented tool toggling via class attributes. Lmk what you think!

sample

File: beaker_kernel/lib/context.py [hunk 7]
```diff
@@ -41,13 +46,25 @@ def __init__(self, beaker_kernel: "LLMKernel", agent_cls: "BaseAgent", config: D
...
-            tools=[],
+            tools=subkernel_tools,
         )
-        self.config = config
-        self.subkernel = self.get_subkernel()
 
+        # Remove tools
+        # NOTE: Identical toolnames don't work
+        enabled_tools = [tool.split(".")[-1] for tool in self.agent.tools.keys()]
+        toggles = [attr[len(TOOL_TOGGLE_PREFIX):].lower() for attr in dir(self) if attr.startswith(TOOL_TOGGLE_PREFIX)]
```
Comment (mattprintz) [inline beaker_kernel/lib/context.py:60]: Since we're pinned to 3.10, we can use [str.removeprefix()](https://docs.python.org/3/library/stdtypes.html#str.removeprefix) instead of subsetting via indexing.

**Comment by fivegrant** [comment]
@mattprintz @brandomr Here's an explanation of the state of tool toggling in this PR.
## With the current code in this PR
As the code currently exists, there are two approaches to disabling the tools in askem-beaker.

The first approach is to include the disabling lines in each context. For example:
```python
class MimiContext(BaseContext):
   TOOL_ENABLE_RUN_CODE = False
   TOOL_ENABLE_ASK_USER = False
   ...
```

The other approach is to make a new context:

```python
class TerariumContext(BaseContext):
   TOOL_ENABLE_RUN_CODE = False
   TOOL_ENABLE_ASK_USER = False
```
and have all of the BaseContexts inherit from it:
```python
class MimiContext(TerariumContext):
   ...
```

## With a change to the code in this PR
If our policy is to avoid making any changes in askem-beaker, we can add in defaults specified by environment variables. Our `.env` would contain lines like
```
DEFAULT_ENABLE_RUN_CODE=true
DEFAULT_ENABLE_ASK_USER=false
```
If there is no environment variable specified, it would default to true. As long as askem-beaker changed its environment variables, it wouldn't need to make any code changes to the contexts. A context could still use a tool despite the default like this:
```python
class MimiContext(BaseContext):
   TOOL_ENABLE_ASK_USER = True # overrides the default specified in the environment
   ...
```

## Picking an approach
Any of these solutions will require some kind of code change. The approaches in the *With the current code in this PR* section expect a change in askem-beaker and *With a change to the code in this PR* section expects a code change in beaker-kernel.

Personally, I prefer creating the new `TerariumContext`. I think the first approach I outlined is messy an error prone. The code is very repetitive for behavior we want to generally apply to each context. As for the default environment change, this new toggling mechanism was created out of a desire to avoid environment variables and complicated, non-discoverable disabling behavior. 

I think we should encourage developers to use a pattern where they create a `class ProjectBaseContext(BaseContext)` that all other contexts inherit from. It's very clear whether a tool is enabled. If the child doesn't explicitly enable/disable, look at the parent. If the parent doesn't have it set either, then it's going to be enabled. I think this is an ideal pattern as opposed to using environment defaults. Environment defaults only seem like a good idea if we're trying to accommodate askem-beaker in the short term.

Another reason to make a `TerariumContext` is we've discussed creating it before for other features (e.g. summarization). We might as well start moving towards growing it out incrementally and this PR seems like a good first step.

Ultimately though I'm fine with any direction we decide to go in

**Comment by mattprintz** [comment]
> a desire to avoid environment variables

Can you expand on this a bit? Is this a desire to avoid all environment variables, or to avoid an explosion of tool/context specific environment variables, etc?

Why is it desirable to avoid environment variables? Particularly since when working with containers, env variables is usually a primary way of setting up configuration options and enabling/disabling behavior in different containers and/or environments (prod, staging, testing, etc)

**Comment by fivegrant** [comment]
@mattprintz 
If we want to vary the defaults based on deployment using the current code, we could always just
```python
class TerariumContext(BaseContext):
   TOOL_ENABLE_RUN_CODE = os.environ.get("DEFAULT_ENABLE_RUN_CODE".lower() == "true", False)
   ...
```

I'm not super against environment variables. It did seem like we were trying to avoid them if possible in previous discussions but I could have misinterpreted.

If we used environment variables, we have something like this
```python
class TerariumContext(BaseContext):
  # Uncomment the line below if you want to override `DEFAULT_ENABLE_RUN_CODE`
  # TOOL_ENABLE_RUN_CODE = True
  ...
```

My preference is still (1) where _you_ make your own environment default mechanism if you want one although I could be nudged to do something like (2). I'm inclined to keep the core functionality bare bones as possible and non-prescriptive as possible.

**Comment by mattprintz** [comment]
> It did seem like we were trying to avoid them if possible in previous discussions but I could have misinterpreted.

What I was getting at with [the earlier comments](https://github.com/jataware/beaker-kernel/pull/47#discussion_r1594546342) is that environment variables are server wide, so a single `ENABLE_USER_PROMPT` environment variable doesn't have enough granularity for what we were looking for. I think I had thrown the idea of out per-context env variables as an option, so we could still use env variables and be granular enough, but that solution does have some flaws as well, so isn't perfect by any stretch.

In short, I never meant to move away from using env variables, just that we shouldn't use a single global env variable that would apply to all contexts.

**Comment by mattprintz** [comment]
I think we are very close here, I just think we need to clean up some of the semantics.
Reading through, I was having trouble figuring out when the tools are available to run and when they aren't, I think I have figured it out now, but I think by renaming things we could make this clearer.

Mostly, I think the confusion is around `TOOL_ENABLE_` as a prefix as to me this seems to imply that the tools referenced are disabled by default and by adding that you are enabling them, but in fact it's the opposite.
TOOL_ENABLE is a verb, implying that by using it, you are doing something, but `TOOL_ENABLE_ASK_USER = True` is (usually) essentially a no-op. As the default is that it is already available. Expecting users to understand that disable a tool they should use `TOOL_ENABLE_foo = False` feels non-intuitive.

There are a couple ways that I see that we can fix this, but I think the most intuitive option is to just switch it from describing an action that should be preformed to describing a state that represents the desired internal state of the context. I.e. to rename from `TOOL_ENABLE_` to `TOOL_ENABLED_`, or something semantically equivalent.

`TOOL_ENABLED_ASK_USER = False` does not read like an English sentence and is awkward, but the semantic meaning is clearer.
`TOOL_ASK_USER_ENABLED = False` reads better, but has an awkward structure with both a prefix and suffix.
Using advanced type hinting we could go with something like:
`TOOL_ASK_USER = 'enabled'` where 'enabled' is defined as [typing.Literal](https://pypi.org/help/#verified-email) choice or even have the values be an enum such as perhaps `TOOL_ASK_USER = ToolOverrdes.ENABLED` or somethng.

I'm open to your input on this.

**Comment by fivegrant** [comment]
@mattprintz what do you think of something like this?

```python3
>>> from enum import Enum
>>> class ToolToggle(Enum):
...     ENABLE = True
...     DISABLE = False
...     def __bool__(self):
...             return self == self.ENABLE
...
>>> # then inside the beaker-kernel lib we check something like
>>> is_enabled_by_dev = ToolToggle.DISABLE
>>> if not is_enabled_by_dev:
...     print("tool disabled")
...
tool disabled
>>>
```

Also, do you think there's a better name than `ToolOverrides` or `ToolToggle`?

File: beaker_kernel/lib/agent.py (hunk 4 above)
Comment (fivegrant) [reply beaker_kernel/lib/agent.py]: this conversation has been moved out into the primary thread. closing this for now

**Comment by mattprintz** [comment]
> @mattprintz what do you think of something like this?> 
> ...
> Also, do you think there's a better name than `ToolOverrides` or `ToolToggle`?

I think the tense is still wrong, and it should be "disabled" and "enabled" as those speak to the state that one want's it to be set which is the appropriate framework for talking about configuration options, which these are.
The __bool__ is redundant, but we shouldn't be checking by value anyway, We would want to refer to/compare against the enums always, not expect it to be a boolean option. 

Your code is semantically no different than what you had before, really, just more complex.

**Review by brandomr — APPROVED** [review]

**Review by brandomr — APPROVED** [review]

**Comment by brandomr** [comment]
Looks good to me. @fivegrant before we merge it can you just add a note somewhere in the docs about how the enabling/disabling works?

**Review by mattprintz — APPROVED** [review]
Looks good pending final tests and documentation.
Thanks for the hard work to get this right Five!

**Comment by fivegrant** [comment]
Sample environment variable changes in darpa-askem are shown here: https://github.com/DARPA-ASKEM/beaker-kernel/tree/feat/upgrade-beaker-kernel-lib.  `ENABLE_USER_PROMPT` needs to be replaced by `TOOL_ENABLE_ASK_USER`. On the other hand, `run_code` isn't enabled by what we currently have anyway (because the checkpointing env variable `ENABLE_CHECKPOINTS` defaults to false) but I went ahead and put the env variables to be explicit

File: docs/contexts.md [hunk 8]
```diff
@@ -147,3 +147,21 @@ A subkernel procedure can be rendered using the `get_code()` method on an
...
 on the context object.
+
+
+## Tool toggling
+
+Tools on the agent can be toggled individually using both environment and 
+class variables. To toggle a tool, create an attribute on your context class or
+a variable in your environment named `TOOL_ENABLED_{YOUR_TOOL_NAME_IN_UPPER_CASE}`.
+'True' enables the tool and 'False' disables it. If the variable is set in
+the environment and class, the value will be taken form the class. For example,
```
Comment (brandomr) [inline docs/contexts.md:158]: from*

**Review by mattprintz — APPROVED** [review]
:+1:
